### PR TITLE
[GRPC_PLUG] Add gRPC connection plugin support 

### PR DESCRIPTION
**Pull request by GomathiselviS**
##### SUMMARY

Fixes https://github.com/ansible/ansible/issues/38820

This PR adds gRPC connection plugin.

Currently it uses ansible_network_os to load the correct

gRPC implementation plugins (will be added in subsequent PR's).

TODO: In future need to add support to work with non-network use case

##### ISSUE TYPE

- Feature Pull Request

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: plugins/connection/grpc.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        (assuming that the secure channel is an SSL channel). If this parameter is
+        specified and the underlying is not an SSL channel, it will just be ignored.
+    ini:
+      - section: grpc_connection
+        key: ssl_target_name_override
+    env:
+      - name: ANSIBLE_GPRC_SSL_TARGET_NAME_OVERRIDE
+    vars:
+      - name: ansible_grpc_ssl_target_name_override
+  persistent_connect_timeout:
```
Comment (ganeshrn) [inline plugins/connection/grpc.py:116]: The option can be referred as an extended doc fragment https://github.com/ansible-collections/ansible.netcommon/blob/main/plugins/connection/network_cli.py#L21

File: plugins/connection/grpc.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        will fail.
+    default: 30
+    ini:
+      - section: persistent_connection
+        key: connect_timeout
+    env:
+      - name: ANSIBLE_PERSISTENT_CONNECT_TIMEOUT
+    vars:
+      - name: ansible_connect_timeout
+  persistent_command_timeout:
```
Comment (ganeshrn) [inline plugins/connection/grpc.py:131]: Same as above

File: plugins/connection/grpc.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+    transport = 'grpc'
+    has_pipelining = False
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
+
+        # TODO: Need to add support to make grpc connection work with non-network target host.
+        # Currently this works only with network target host.
+        if self._network_os:
```
Comment (ganeshrn) [inline plugins/connection/grpc.py:204]: iiuc the use for `grpc_type` option I think it should also check it user has set the `grpc_type` configuration option with a `or` condition.

File: plugins/connection/grpc.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+            cref = dict(zip(["corg", "cname", "plugin"], grpc_type.split(".")))
+            grpclib = "ansible_collections.{corg}.{cname}.plugins.sub_plugins.grpc.{plugin}".format(
+                **cref
+            )
+            grpccls = getattr(import_module(grpclib), "Grpc")
+            grpc_obj = grpccls(self)
+
+            if grpc_obj:
+                self._sub_plugin = {'type': 'grpc', 'name': grpc_type, 'obj': grpc_obj}
+                self.queue_message('log', 'loaded gRPC plugin for network_os %s' % self._network_os)
```
Comment (ganeshrn) [inline plugins/connection/grpc.py:220]: This can be rephrased to `loaded gRPC plugin for type`

File: plugins/connection/grpc.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+                self.queue_message('log', 'network_os is set to %s' % self._network_os)
```
Comment (ganeshrn) [inline plugins/connection/grpc.py:221]: `network_os is set to %s` can be changed to `grpc type is set to %s`

File: plugins/connection/grpc.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+            else:
+                raise AnsibleConnectionFailure('unable to load API plugin for network_os %s' % self._network_os)
+        else:
+            self._sub_plugin['type'] = 'external'
+            self.queue_message('warning', 'Unable to automatically determine gRPC implementation type.'
+                                          ' Please manually configure ansible_network_os value for this host')
```
Comment (ganeshrn) [inline plugins/connection/grpc.py:227]: Can update the error message to mention `grpc_type` as well, `Please manually configure ansible_network_os value or grpc_type configuration variable for this host`

File: plugins/module_utils/network/grpc/grpc.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
+# (c) 2019 Red Hat, Inc.
```
Comment (ganeshrn) [inline plugins/module_utils/network/grpc/grpc.py:1]: Please update the year

File: plugins/module_utils/network/grpc/grpc.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        if out:
+            err = json.loads(out)
+            res = json.dumps(err, indent=4, separators=(',', ': '))
+    except ValueError as err:
+        if check_rc:
+            module.fail_json(msg=to_text(out['error'], errors='surrogate_then_replace'))
+        else:
+            module.warn(to_text(out['error'], errors='surrogate_then_replace'))
+        return err
+    return
```
Comment (ganeshrn) [inline plugins/module_utils/network/grpc/grpc.py:74]: This return statement is not required

File: plugins/module_utils/network/grpc/grpc.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            module.warn(to_text(out['error'], errors='surrogate_then_replace'))
+
+    return response.strip(), error.strip()
+
+def merge_config(module, section, check_rc=True):
+    conn = get_connection(module)
+    try:
+        out = conn.merge_config(section)
+        if out:
+            err = json.loads(out)
```
Comment (ganeshrn) [inline plugins/module_utils/network/grpc/grpc.py:66]: Where is `err` and `res` used if there is not expection

File: plugins/module_utils/network/grpc/grpc.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    conn = get_connection(module)
+    try:
+        out = conn.replace_config(section)
+        if out:
+            err = json.loads(out)
+            res = json.dumps(err, indent=4, separators=(',', ': '))
+            module.fail_json(msg=to_text(res, errors='surrogate_then_replace'))
+    except AbortionError:
+        module.fail_json(msg=to_text("Unable to connect to grpc destination", errors='surrogate_then_replace'))
+    return
```
Comment (ganeshrn) [inline plugins/module_utils/network/grpc/grpc.py:86]: Same as above

File: plugins/module_utils/network/grpc/grpc.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+def delete_config(module, section, check_rc=True):
+    conn = get_connection(module)
+    try:
+        out = conn.delete_config(section)
+        if out and out:
```
Comment (ganeshrn) [inline plugins/module_utils/network/grpc/grpc.py:92]: ` if out and out:` can be replaced with ` if out:`

File: plugins/module_utils/network/grpc/grpc.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+            err = json.loads(out)
+            res = json.dumps(err, indent=4, separators=(',', ': '))
+    except ValueError as err:
+        if check_rc:
+            module.fail_json(msg=to_text(out['error'], errors='surrogate_then_replace'))
+        else:
+            module.warn(to_text(out['error'], errors='surrogate_then_replace'))
+        return err
+    return
```
Comment (ganeshrn) [inline plugins/module_utils/network/grpc/grpc.py:101]: Same as above

File: plugins/modules/grpc_nw_config.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+  state:
+    description: action to be performed
+   
+requirements:
+  - grpcio
+  - protobuf
+notes:
+  - This module requires the gRPC system service be enabled on
+    the target host being managed.
+  - This module supports the use of connection=grpc.
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_config.py:39]: `connection=grpc` to `connection=ansible.netcommon.grpc`

File: plugins/modules/grpc_nw_config.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+  - This module requires the value of 'ansible_network_os' be defined as an inventory variable.
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_config.py:40]: Can be rephrased to `This module requires the value of 'ansible_network_os' or `grpc_type` configuration option (refer `ansible.netcommon.grpc` connection plugin documentation) be defined as an inventory variable.`

File: plugins/modules/grpc_nw_config.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+    - gRPC is a high performance, open-source universal RPC framework.
+    - This module allows the user to append configs to an existing configuration in a gRPC
+      enabled devices.
+options:
+  config:
+    description:
+      - This option specifies the string which acts as a filter to restrict the portions of
+        the data to be retrieved from the target host device. If this option is not specified the entire
+        configuration or state data is returned in response provided it is supported by target host.
+  path:
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_config.py:28]: I think it is better to have option as `config` which accepts the input config in json format instead of file `path` as user can read from file using file lookup plugin or provide values from `host_vars`

File: plugins/modules/grpc_nw_config.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+notes:
+  - This module requires the gRPC system service be enabled on
+    the target host being managed.
+  - This module supports the use of connection=grpc.
+  - This module requires the value of 'ansible_network_os' be defined as an inventory variable.
+  - Tested against iosxrv 9k version 6.1.2.
+"""
+
+EXAMPLES = """
+- name: Merge bgp configuration data
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_config.py:45]: Can add example with `config` option with value provided in the task and another example of value read from file using the `file` lookup plugin.

File: plugins/modules/grpc_nw_config.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+  grpc_nw_config:
+    path:  'bgp_merge.json'
+    state: merged
+"""
+
+RETURN = """
+stdout:
+  description: The raw string containing response object
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_config.py:53]: This can be update to empty string if no response from server.

File: plugins/modules/grpc_nw_config.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+  config:
+    description:
+      - This option specifies the string which acts as a filter to restrict the portions of
+        the data to be retrieved from the target host device. If this option is not specified the entire
+        configuration or state data is returned in response provided it is supported by target host.
+  path:
+    description: path of the file that has the config in json format.
+  state:
+    description: action to be performed
+
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_config.py:32]: Can add a `backup` option to create a back file similar to other config modules.

File: plugins/modules/grpc_nw_get.py [hunk 19]
```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright 2020 Red Hat
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_get.py:3]: Update the year

File: plugins/modules/grpc_nw_get.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+        etc. based on what is supported by the gRPC server. By default it will return both configuration
+        and operational state data in response.
+requirements:
+  - grpcio
+  - protobuf
+notes:
+  - This module requires the gRPC system service be enabled on
+    the target host being managed.
+  - This module supports the use of connection=grpc.
+  - This module requires the value of 'ansible_network_os' be defined as an inventory variable.
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_get.py:52]: Same as above

File: plugins/modules/grpc_nw_get.py [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+        capability of the gRPC server running on target host. The values can be I(config), I(oper)
+        etc. based on what is supported by the gRPC server. By default it will return both configuration
+        and operational state data in response.
+requirements:
+  - grpcio
+  - protobuf
+notes:
+  - This module requires the gRPC system service be enabled on
+    the target host being managed.
+  - This module supports the use of connection=grpc.
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_get.py:51]: Same as above

File: plugins/modules/grpc_nw_get.py [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+author:
+    - "Ganesh Nalawade (@ganeshrn)"
+    - "Gomathi Selvi S (@GomathiselviS)
+short_description: Fetch configuration/state data from gRPC enabled target hosts.
+description:
+    - gRPC is a high performance, open-source universal RPC framework.
+    - This module allows the user to fetch configuration and state data from gRPC
+      enabled devices.
+options:
+  section:
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_get.py:24]: We can add a alias to this option called `filter`

File: plugins/modules/grpc_nw_get.py [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+    if module.params['section']:
+        section = json.dumps(yaml.safe_load(module.params['section']))
+    else:
+        section = None
+    command = module.params['command']
+    display = module.params['display']
+    data_type = module.params['data_type']
+
+    if display and display not in operations.get('display', []):
+        module.fail_json(msg="display option '%s' is not supported on this target host" % display)
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_get.py:160]: Please add information about supported display type.
```
    supported_display = operations.get('display', []
    if display and display not in supported_display):
        module.fail_json(msg="display option '%s' is not supported on this target host. Valid value is one of %s" % display, supported_display.join(", "))
```

File: plugins/modules/grpc_nw_get.py [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+
+    if command and not operations.get('supports_cli_command', False):
+        module.fail_json(msg="command option '%s' is not supported on this target host" % command)
+
+    if data_type and data_type not in operations.get('data_type', []):
+        module.fail_json(msg="data_type option '%s' is not supported on this target host" % data_type)
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_get.py:166]: Add information about supported data type in error message same as above.

File: plugins/modules/grpc_nw_get.py [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+
+    result = {'changed': False}
+    output = []
+
+    try:
+        if command:
+            response, err = run_cli(module, command, display)
+        else:
+            response, err = get(module, section, data_type)
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_get.py:175]: How the value of err is handled here?

File: plugins/modules/grpc_nw_get.py [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+
+        try:
+            output = json.loads(response)
+        except ValueError:
+            output = None
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_get.py:180]: Why is output set to `None` here?

File: plugins/modules/grpc_nw_get.py [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+    except ConnectionError as exc:
+        module.fail_json(msg=to_text(exc, errors='surrogate_then_replace'), code=exc.code)
+    
+
+    #result['stdout'] = yaml.safe_dump(response)
```
Comment (ganeshrn) [inline plugins/modules/grpc_nw_get.py:185]: Can remove this line

File: plugins/sub_plugins/grpc/base.py [hunk 28]
```diff
@@ -0,0 +1,63 @@
+# (c) 2019 Ansible Project
```
Comment (ganeshrn) [inline plugins/sub_plugins/grpc/base.py:1]: Update the year

File: plugins/connection/grpc.py [hunk 29]
```diff
@@ -0,0 +1,292 @@
...
+    type: str
+    description: grpc type
+    default: False
+    ini:
+      - section: grpc_connection
+        key: type
+    env:
+      - name: ANSIBLE_GRPC_CONNECTION_TYPE
+    vars:
+      - name: ansible_grpc_connection_type
```
[the remark this answers is not in the document]
Comment (GomathiselviS) [reply plugins/connection/grpc.py:172]: @Qalthos I have updated the description for this option. This can be used instead of network_os.

File: plugins/connection/grpc.py (hunk 1 above)
Comment (GomathiselviS) [reply plugins/connection/grpc.py]: Done

File: plugins/connection/grpc.py (hunk 2 above)
Comment (GomathiselviS) [reply plugins/connection/grpc.py]: Done

File: plugins/connection/grpc.py [hunk 30]
```diff
@@ -0,0 +1,292 @@
...
+        closed.
+    default: 30
+    ini:
+      - section: persistent_connection
+        key: command_timeout
+    env:
+      - name: ANSIBLE_PERSISTENT_COMMAND_TIMEOUT
+    vars:
+      - name: ansible_command_timeout
+  persistent_log_messages:
```
[the remark this answers is not in the document]
Comment (GomathiselviS) [reply plugins/connection/grpc.py]: Done

File: plugins/action/grpc_config.py [hunk 31]
```diff
@@ -0,0 +1,90 @@
...
+        warnings = []
+
+        if persistent_connection != "grpc" and module_name == "grpc_nw_config":
+            return {
+                "failed": True,
+                "msg": "Connection type %s is not valid for netconf_config module. "
+                "Valid connection type is grpc"
+                % self._play_context.connection,
+            }
+        if module_name == "grpc_nw_config":
```
Comment (NilashishC) [inline plugins/action/grpc_config.py:52]: Why is this part required?

File: plugins/action/grpc_config.py [hunk 32]
```diff
@@ -0,0 +1,90 @@
...
+
+from ansible.utils.display import Display
+from ansible_collections.ansible.netcommon.plugins.action.network import (
+    ActionModule as ActionNetworkModule,
+)
+
+display = Display()
+
+
+class ActionModule(ActionNetworkModule):
```
Comment (NilashishC) [inline plugins/action/grpc_config.py:34]: Since the action plugins are picked up based on the module prefix, REF: https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_executor.py#L1109-L1119, can we not name the action plugin just grpc.py?

File: plugins/action/grpc_config.py [hunk 33]
```diff
@@ -0,0 +1,90 @@
...
+        del tmp  # tmp no longer has any effect
+
+        module_name = self._task.action.split(".")[-1]
+        self._config_module = (
+            True if module_name == "grpc_nw_config" else False
+        )
+        persistent_connection = self._play_context.connection.split(".")[-1]
+        warnings = []
+
+        if persistent_connection != "grpc" and module_name == "grpc_nw_config":
```
Comment (NilashishC) [inline plugins/action/grpc_config.py:45]: I do not see a grpc_nw_config module in this PR. Am I missing something?

File: plugins/action/grpc_config.py [hunk 34]
```diff
@@ -0,0 +1,90 @@
...
+                "failed": True,
+                "msg": "Connection type %s is not valid for netconf_config module. "
+                "Valid connection type is grpc"
+                % self._play_context.connection,
+            }
+        if module_name == "grpc_nw_config":
+            args = self._task.args
+            pc = copy.deepcopy(self._play_context)
+            pc.connection = "ansible.netcommon.grpc"
+            pc.port = int(args.get("port") or self._play_context.port or 57777)
```
Comment (NilashishC) [inline plugins/action/grpc_config.py:56]: Would task args ever contain any inventory related data for plugins that do not have provider or top spec?

File: plugins/action/grpc_config.py [hunk 35]
```diff
@@ -0,0 +1,90 @@
+#
+# Copyright 2018 Red Hat Inc.
```
Comment (NilashishC) [inline plugins/action/grpc_config.py:2]:
```suggestion
# Copyright 2022 Red Hat Inc.
```

File: plugins/connection/grpc.py [hunk 36]
```diff
@@ -0,0 +1,294 @@
...
+    ini:
+      - section: grpc_connection
+        key: ssl_target_name_override
+    env:
+      - name: ANSIBLE_GPRC_SSL_TARGET_NAME_OVERRIDE
+    vars:
+      - name: ansible_grpc_ssl_target_name_override
+  grpc_type:
+    description:
+        - This option indicated the grpc type and it can be used
```
Comment (NilashishC) [inline plugins/connection/grpc.py:121]: What is the role of grpc_type parameter?

File: plugins/connection/grpc.py [hunk 37]
```diff
@@ -0,0 +1,292 @@
...
+    ini:
+      - section: persistent_connection
+        key: log_messages
+    env:
+      - name: ANSIBLE_PERSISTENT_LOG_MESSAGES
+    vars:
+      - name: ansible_persistent_log_messages
+  grpc_type:
+    type: str
+    description: grpc type
```
[the remark this answers is not in the document]
Comment (GomathiselviS) [reply plugins/connection/grpc.py]: Done

File: plugins/modules/grpc_nw_config.py (hunk 15 above)
Comment (GomathiselviS) [reply plugins/modules/grpc_nw_config.py]: Done

File: plugins/action/grpc_config.py (hunk 33 above)
Comment (GomathiselviS) [reply plugins/action/grpc_config.py:45]: I renamed the module but forgot to update it here. I will fix it.

File: plugins/action/grpc_config.py [hunk 38]
```diff
@@ -0,0 +1,48 @@
...
+class ActionModule(ActionNetworkModule):
+    def run(self, tmp=None, task_vars=None):
+        del tmp  # tmp no longer has any effect
+
+        self._config_module = True
+        warnings = []
+        if self._play_context.connection.split(".")[-1] != "grpc":
+            return {
+                "failed": True,
+                "msg": "Connection type %s is not valid for netconf_config module. "
```
Comment (NilashishC) [inline plugins/action/grpc_config.py:37]:
```suggestion
                "msg": "Connection type %s is not valid for grpc_ module. "
```

**Comment by GomathiselviS** [comment]
recheck

File: plugins/connection/grpc.py [hunk 39]
```diff
@@ -0,0 +1,294 @@
...
+
+        # TODO: Need to add support to make grpc connection work with non-network target host.
+        # Currently this works only with network target host.
+        grpc_type = self._network_os or self.get_option("grpc_type")
+        if grpc_type:
+            if not HAS_PROTOBUF:
+                raise AnsibleError(
+                    "protobuf is required to use the grpc connection type. Please run 'pip install protobuf'"
+                )
+            if self._network_os:
```
Comment (ganeshrn) [inline plugins/connection/grpc.py:174]: replace `self._network_os` with `grpc_type` wherever applicable

File: plugins/modules/grpc_get.py [hunk 40]
```diff
@@ -0,0 +1,220 @@
...
+        and operational state data in response.
+    type: str
+requirements:
+  - grpcio
+  - protobuf
+notes:
+  - This module requires the gRPC system service be enabled on
+    the target host being managed.
+  - This module supports the use of connection=ansible.netcommon.grpc.
+  - This module requires the value of 'ansible_network_os' configuration option (referansible.netcommon.grpc
```
Comment (ganeshrn) [inline plugins/modules/grpc_get.py:58]: can be updated to reflect `grpc_type`
`(referansible.netcommon.grpc connection plugin documentation)` with `(refer ansible.netcommon.grpc connection plugin documentation)`

File: plugins/connection/grpc.py (hunk 39 above)
Comment (GomathiselviS) [reply plugins/connection/grpc.py:174]: This if condition is for the case when grpc_type is not given and network_os is given. From the network_os, grpc_type will be formed.

File: plugins/action/grpc_config.py [hunk 41]
```diff
@@ -0,0 +1,48 @@
...
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+from ansible_collections.ansible.netcommon.plugins.action.network import (
+    ActionModule as ActionNetworkModule,
+)
+
+
+class ActionModule(ActionNetworkModule):
```
Comment (NilashishC) [inline plugins/action/grpc_config.py:28]: I think it's better to rename this to `action/grpc.py` so that all modules prefixed with `grpc_` can benefit from the connection type check.

File: plugins/modules/grpc_nw_get.py (hunk 25 above)
Comment (GomathiselviS) [reply plugins/modules/grpc_nw_get.py]: Fixed this.

File: plugins/modules/grpc_nw_get.py (hunk 26 above)
Comment (GomathiselviS) [reply plugins/modules/grpc_nw_get.py]: Removed this.

File: plugins/action/grpc_config.py (hunk 31 above)
Comment (GomathiselviS) [reply plugins/action/grpc_config.py]: These lines are removed now.

File: plugins/action/grpc_config.py (hunk 34 above)
Comment (GomathiselviS) [reply plugins/action/grpc_config.py]: Removed these lines.

File: plugins/action/grpc.py [hunk 42]
```diff
@@ -0,0 +1,48 @@
...
+class ActionModule(ActionNetworkModule):
+    def run(self, tmp=None, task_vars=None):
+        del tmp  # tmp no longer has any effect
+
+        self._config_module = True
+        warnings = []
+        if self._play_context.connection.split(".")[-1] != "grpc":
+            return {
+                "failed": True,
+                "msg": "Connection type %s is not valid for grpc_config module. "
```
Comment (NilashishC) [inline plugins/action/grpc.py:37]: Minor comment - the module name can be generic here.

File: plugins/action/grpc_config.py [hunk 43]
```diff
@@ -0,0 +1,48 @@
...
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
```
Comment (NilashishC) [inline plugins/action/grpc_config.py:17]: If we already have a grpc.py action plugin, what is the requirement of this file?

File: plugins/module_utils/network/grpc/grpc.py [hunk 44]
```diff
@@ -0,0 +1,156 @@
...
+def get_connection(module):
+    if hasattr(module, "_grpc_connection"):
+        return module._grpc_connection
+
+    capabilities = get_capabilities(module)
+    network_api = capabilities.get("network_api")
+    if network_api == "ansible.netcommon.grpc":
+        module._grpc_connection = Connection(module._socket_path)
+    else:
+        module.fail_json(msg="Invalid connection type %s" % network_api)
```
Comment (NilashishC) [inline plugins/module_utils/network/grpc/grpc.py:39]: Is this still needed if we're already checking this in the action plugin?

File: plugins/action/grpc_config.py [hunk 45]
```diff
@@ -0,0 +1,51 @@
...
+__metaclass__ = type
+
+from ansible_collections.ansible.netcommon.plugins.action.network import (
+    ActionModule as ActionNetworkModule,
+)
+
+
+class ActionModule(ActionNetworkModule):
+    def run(self, tmp=None, task_vars=None):
+        import q
```
Comment (ganeshrn) [inline plugins/action/grpc_config.py:30]: Please remove `q` and related statements

File: plugins/connection/grpc.py (hunk 39 above)
Comment (ganeshrn) [reply plugins/connection/grpc.py:174]: `grpc_type` will be FQCN right? In that case the `if else` clause is not required. IIUC the user can set either `grpc_type` or `ansible_network_os` to `cisco.iosxr.grpc`

File: plugins/connection/grpc.py [hunk 46]
```diff
@@ -0,0 +1,294 @@
...
+    transport = "ansible.netcommon.grpc"
+    has_pipelining = False
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+
+        super(Connection, self).__init__(
+            play_context, new_stdin, *args, **kwargs
+        )
+
+        # TODO: Need to add support to make grpc connection work with non-network target host.
```
Comment (ganeshrn) [inline plugins/connection/grpc.py:166]: This comment can be removed as we have added `grpc_type` plugin config option

File: plugins/module_utils/network/grpc/grpc.py (hunk 44 above)
Comment (ganeshrn) [reply plugins/module_utils/network/grpc/grpc.py:39]: action plugin will be invoked only for `grpc_config` module as the name of the action plugin file is `grpc_config.py` 
The name of the action plugin file can be renamed to `grpc.py` and redirection for `grpc_config` and `grpc_get` modules can be added to the action plugin in `runtime.yml` file.

File: plugins/action/grpc_config.py (hunk 43 above)
Comment (GomathiselviS) [reply plugins/action/grpc_config.py:17]: grpc.py is not getting picked up

File: plugins/module_utils/network/grpc/grpc.py (hunk 44 above)
Comment (GomathiselviS) [reply plugins/module_utils/network/grpc/grpc.py:39]: Removing grpc.py for now

Comment (GomathiselviS) [reply plugins/module_utils/network/grpc/grpc.py:39]: Done

File: plugins/sub_plugins/grpc/base.py [hunk 47]
```diff
@@ -0,0 +1,65 @@
...
+    A base class for implementing gRPC abstraction layer
+    """
+
+    __rpc__ = ["channel", "get_config", "edit_config", "get"]
+
+    def __init__(self, connection):
+        super(GrpcBase, self).__init__()
+        self._connection = connection
+
+    @property
```
Comment (Qalthos) [inline plugins/sub_plugins/grpc/base.py:35]:
```suggestion
    @property
    @ensure_connect
```

iosxr is the only implementation linked, and it doesn't appear to use this property, but it looks like you need to be connected for this, and if we always go through `self.channel` we shouldn't need ensure_connect in the implementation

File: plugins/connection/grpc.py [hunk 48]
```diff
@@ -0,0 +1,292 @@
...
+  - Ansible Networking Team (@ansible-network)
+name: grpc
+short_description: Provides a persistent connection using the gRPC protocol
+description:
+  - This connection plugin provides a connection to remote devices over gRPC and
+    is typically used with devices for sending and receiving RPC calls
+    over gRPC framework.
+  - Note this connection plugin requires the grpcio python library to be installed on the
+    local Ansible controller.
+version_added: "3.0.0"
```
Comment (Qalthos) [inline plugins/connection/grpc.py:18]:
```suggestion
version_added: "3.1.0"
```

File: plugins/modules/grpc_config.py [hunk 49]
```diff
@@ -0,0 +1,248 @@
...
+# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+module: grpc_config
+version_added: "3.0.0"
```
Comment (Qalthos) [inline plugins/modules/grpc_config.py:14]:
```suggestion
version_added: "3.1.0"
```

File: plugins/connection/grpc.py [hunk 50]
```diff
@@ -0,0 +1,292 @@
...
+                    "unable to load API plugin for network_os %s"
+                    % self._network_os
+                )
+        else:
+            self._sub_plugin["type"] = "external"
+            self.queue_message(
+                "warning",
+                "Unable to automatically determine gRPC implementation type."
+                " Please manually configure ansible_network_os value or grpc_type configuration for this host",
+            )
```
Comment (Qalthos) [inline plugins/connection/grpc.py:208]: Why isn't this a failure? `type: external` is for when the object in `obj` can't be loaded by PluginLoader, but that's not really the case here.

**Review by Qalthos — CHANGES_REQUESTED** [review]

File: plugins/connection/grpc.py (hunk 50 above)
Comment (GomathiselviS) [reply plugins/connection/grpc.py:208]: Agreed. Fixing it.

File: plugins/connection/grpc.py (hunk 39 above)
Comment (GomathiselviS) [reply plugins/connection/grpc.py:172]: Is it not expected from the user to set ansible_network_os as 'cisco.iosxr.iosxr' ?

File: docs/ansible.netcommon.cli_command_module.rst [hunk 51]
```diff
@@ -1,268 +0,0 @@
-.. _ansible.netcommon.cli_command_module:
```
Comment (ganeshrn) [inline docs/ansible.netcommon.cli_command_module.rst:1]: Why are the docs removed in this PR?

File: tests/integration/targets/grpc_config/tasks/main.yaml [hunk 52]
```diff
@@ -0,0 +1,15 @@
...
+
+- include: iosxr.yaml
+  when: ansible_network_os == 'cisco.iosxr.iosxr'
+  tags:
+    - netconf
+
+- include: nxos.yaml
+  when: ansible_network_os == 'cisco.nxos.nxos'
+  tags:
+    - netconf
```
Comment (Qalthos) [inline tests/integration/targets/grpc_config/tasks/main.yaml:15]: Are we going to go back and fill these in later, or is this just going to confuse people?

I'd also rather see these tagged grpc

**Comment by Qalthos** [comment]
recheck

File: tests/integration/targets/grpc_config/tasks/main.yaml (hunk 52 above)
Comment (GomathiselviS) [reply tests/integration/targets/grpc_config/tasks/main.yaml]: This was a mistake. Fixed it.

**Comment by GomathiselviS** [comment]
recheck

**Review by NilashishC — CHANGES_REQUESTED** [review]
Looks like a LOT of doc RST files were deleted. Something might have went wrong while running collection_prep.

**Review by NilashishC — APPROVED** [review]

**Review by Qalthos — APPROVED** [review]

**Review by ganeshrn — APPROVED** [review]
